extensions/ql-vscode/src/view/results/ResultTablesHeader.tsx

```diff
@@ -0,0 +1,141 @@
+import * as React from "react";
+import { useCallback, useEffect } from "react";
+import { vscode } from "../vscode-api";
+import { openFile, tableHeaderItemClassName } from "./result-table-utils";
+import { sendTelemetry } from "../common/telemetry";
+import {
+  ALERTS_TABLE_NAME,
+  ParsedResultSets,
+} from "../../common/interface-types";
+import { basename } from "../../common/path";
+
+interface Props {
+  queryName: string;
+  queryPath: string;
+  parsedResultSets: ParsedResultSets;
+  selectedTable: string;
+}
+
+export function ResultTablesHeader(props: Props) {
+  const { queryPath, queryName, parsedResultSets, selectedTable } = props;
+
+  const [selectedPage, setSelectedPage] = React.useState(
+    `${parsedResultSets.pageNumber + 1}`,
+  );
+  useEffect(() => {
+    setSelectedPage(`${parsedResultSets.pageNumber + 1}`);
+  }, [parsedResultSets.pageNumber]);
+
+  // FIXME: The extension, not the view, should be in charge of deciding whether to initially show
+  // a raw or alerts page. We have to conditionally recompute the number of pages here, because
+  // on initial load of query results, resultSets.numPages will have the number of *raw* pages available,
+  // not interpreted pages, because the extension doesn't know the view will default to showing alerts
+  // instead.
+  const numPages = Math.max(
+    selectedTable === ALERTS_TABLE_NAME
+      ? parsedResultSets.numInterpretedPages
+      : parsedResultSets.numPages,
+    1,
+  );
+
+  const onChangeHandler = useCallback(
+    (e: React.ChangeEvent<HTMLInputElement>) => {
+      setSelectedPage(e.target.value);
+      sendResultsPageChangedTelemetry();
+    },
+    [],
+  );
+
+  const changePage = useCallback(
+    (value: string) => {
+      const pageNumber = parseInt(value);
+      if (pageNumber !== undefined && !isNaN(pageNumber)) {
+        const actualPageNumber = Math.max(
+          0,
+          Math.min(pageNumber - 1, numPages - 1),
+        );
+        vscode.postMessage({
+          t: "changePage",
+          pageNumber: actualPageNumber,
+          selectedTable,
+        });
+      }
+    },
+    [numPages, selectedTable],
+  );
+
+  const onBlurHandler = useCallback(
+    (e: React.FocusEvent<HTMLInputElement, Element>) => {
+      changePage(e.target.value);
+    },
+    [changePage],
+  );
+
+  const onKeyDownHandler = useCallback(
+    (e: React.KeyboardEvent<HTMLInputElement>) => {
+      changePage(e.currentTarget.value);
```

Comment:
Oops, I forgot the logic about checking if it's keycode 13. I'll add that back in.

Reply:
I've actually done `e.key === "Enter"` because `e.keyCode` is deprecated now. See https://www.w3.org/TR/uievents-key/#named-key-attribute-values for a list of possible values.